src/components/Camera.ts

```diff
@@ -0,0 +1,232 @@
+import { CSSProperties, Component, ReactElement, createElement } from "react";
+import { snap, stream } from "webcamjs";
+
+import { Alert, AlertProps } from "./Alert";
+import { CameraButton } from "./CameraButton";
+import { WebCam } from "./WebCam";
+import { Image } from "./Image";
+
+import "../ui/Camera.scss";
+
+export type CaptionType = "icons" | "buttons";
+
+export interface CameraProps {
+    captureButton: string;
+    captureIcon: string;
+    caption: CaptionType;
+    fileType: string;
+    filter: string;
+    width: number;
+    widthUnit: string;
+    onClickAction: (image: {src: string, id: string }) => void;
+    recaptureButton: string;
+    ref: (camContainer: HTMLDivElement) => void;
+    style?: object;
+    switchCameraIcon: string;
+    savePictureIcon: string;
+    savePictureButton: string;
+    height: number;
+    heightUnit: string;
+}
+
+export interface CameraState {
+    availableDevices: string[];
+    cameraDevicePosition: number;
+    browserSupport: boolean;
+    swapCamera: boolean;
+    pictureId: string;
+    pictureTaken: boolean;
+    retakePhoto: boolean;
+    screenshot: string;
+}
+
+export type FileFormats = "jpeg" | "png" | "webp";
+
+export class Camera extends Component<CameraProps, CameraState> {
+    private webcam?: HTMLDivElement;
+    private availableDevices: string[];
+    private videoElement ?: HTMLVideoElement;
+
+    constructor(props: CameraProps) {
+        super(props);
+
+        this.state = {
+            availableDevices: [],
+            browserSupport: false,
+            cameraDevicePosition: 0,
+            pictureId: "",
+            pictureTaken: false,
+            retakePhoto: false,
+            screenshot: "",
+            swapCamera: false
+        };
+
+        this.availableDevices = [];
```

Comment:
What is the difference between this variable and the one in the state? It is confusing.

Reply:
It was a means for me to do basic unit test on that section of code, ie To allow me set a constrain of multiple devices such that a certain section of the code is tested. I have made an improvement to it though.